Trade history, recent trades, live trade processing and storage

gloriousCode opened the pull request:
# PR Description
This PR adds a way for GCT to process and store trades. The main features include:

- New database table `trade` to store trades
- Implementation of wrapper functions `GetRecentTrades` and `GetHistoricTrades` where applicable
- Live trade processing via the websocket
- Conversion of trades to candles at the intervals you request
  - Allowing the ability to generate candles for periods where exchange's don't natively support them via the API ie 1minute
- Ability to overwrite existing candle data using trade data/API candle data
- New exchange level config variable `saveTradeData`. It is there so that just because you have a database enabled and you want to track trades, doesn't mean you want all exchange's trades
- Management and interactivity of all the above via CLI

To understand how the trade system works, I have a fancy [readme here](https://github.com/gloriousCode/gocryptotrader/tree/trade-spade-escapade/exchanges/trade)


## Type of change
- [✔️] New feature (non-breaking change which adds functionality)
- [✔️] Just some minor bug fixes I came across throughout development

## How has this been tested
- [✔️] go test ./... -race
- [✔️] golangci-lint run
- [✔️] So many tests everywhere
- [✔️] `dbmigrate.exe -command=up` and `dbmigrate.exe -command=down` to ensure happy conversion

## Checklist
- [✔️] My code follows the style guidelines of this project
- [✔️] I have performed a self-review of my own code
- [✔️] I have commented my code, particularly in hard-to-understand areas
- [✔️] I have made corresponding changes to the documentation and regenerated documentation via the documentation tool
- [✔️] My changes generate no new warnings
- [✔️] I have added tests that prove my fix is effective or that my feature works
- [✔️] New and existing unit tests pass locally and on Travis with my changes
- [✔️] Any dependent changes have been merged and published in downstream modules

shazbert reviewed (changes requested):
Very well done! This is a great upgrade. 

Had an issue with coinbene and writing to a psql db:
Coinbene websocket run:
[DEBUG]| 17/09/2020 13:59:44 |SQL: SELECT * FROM "exchange" WHERE (name= $1) LIMIT 1;
[DEBUG]| 17/09/2020 13:59:44 |SQL: [coinbene]
[ERROR]| 17/09/2020 13:59:44 |exchange not found
No idea what that means

Just some basic nits below, you added in some epic code!

gloriousCode commented on database/repository/trade/trade.go:
```diff
@@ -0,0 +1,345 @@
+	}
+	defer func() {
+		if err != nil {
+			err = tx.Rollback()
```
No, its really weird/fun in that way. I tested it out, the caller still receives the original error and this is processed after the function has returned. I can see it being a readability issue though, so I'm happy to change this to a new error

shazbert replied:
Ahh right, things you learn. 👍

gloriousCode commented on exchanges/poloniex/poloniex_websocket.go:
```diff
@@ -340,28 +344,31 @@ func (p *Poloniex) wsHandleData(respRaw []byte) error {
 
-						p.Websocket.DataHandler <- stream.TradeData{
-							Timestamp:    time.Unix(trade.Timestamp, 0),
+						return trade.AddTradesToBuffer(p.Name, trade.Data{
```
Sounds good! I like it

gloriousCode commented on exchanges/bithumb/bithumb_wrapper.go:
```diff
@@ -329,9 +330,50 @@ func (b *Bithumb) GetFundingHistory() ([]exchange.FundHistory, error) {
+			return nil, err
+		}
+		var t time.Time
+		t, err = time.Parse("2006-01-02 15:04:05", tradeData.Data[i].TransactionDate)
```
On the other side of things, some websockets support sending trade snapshots, so you can easily end up inserting duplicated (or run `GetRecentTrades` repeatedly). So some constraint is necessary. I'll keep thinking as well 🧙

shazbert replied:
Yeah, you make a good point. 👍

gloriousCode commented on exchanges/lbank/lbank_wrapper.go:
```diff
@@ -326,9 +327,48 @@ func (l *Lbank) GetFundingHistory() ([]exchange.FundHistory, error) {
+
+// GetHistoricTrades returns historic trade data within the timeframe provided
+func (l *Lbank) GetHistoricTrades(p currency.Pair, assetType asset.Item, timestampStart, timestampEnd time.Time) ([]trade.Data, error) {
+	if _, ok := l.CurrencyPairs.Pairs[assetType]; !ok {
```
Do you mean add something like
```
if timestampEnd.After(time.Now()) {
		return nil, fmt.Errorf("invalid end date supplied '%v'", timestampEnd)
	}
```
to all implemented `GetHistoricTrades`?

shazbert replied:
Yessir!

shazbert reviewed (changes requested):
Super spam
```
gocryptotrader\cmd\gctcli> go run ./... findmissingsavedtradeintervals bitmex eth-usd perpetualcontract
{
 "missingPeriods": [
  "1 hour gap found in trade data for bitmex perpetualcontract eth-usd between between 2020-09-16 14:28:02 and 2020-09-16 15:28:02",
  "1 hour gap found in trade data for bitmex perpetualcontract eth-usd between between 2020-09-16 15:28:02 and 2020-09-16 16:28:02",
  "1 hour gap found in trade data for bitmex perpetualcontract eth-usd between between 2020-09-16 16:28:02 and 2020-09-16 17:28:02",
  "1 hour gap found in trade data for bitmex perpetualcontract eth-usd between between 2020-09-16 17:28:02 and 2020-09-16 18:28:02",
"foundPeriods": [
spam spam spam
```
go run ./... findmissingsavedcandleintervals bitmex eth-usd perpetualcontract 15
2020/09/17 15:25:42 rpc error: code = ResourceExhausted desc = grpc: received message larger than max (17677306 vs. 4194304)

Even more spammy.

Other than that Postgres was tested Sqlite3 was tested. Dbmigrated up and down 👍.  Everything looks to be inserted correctly. 

Out of scope considerations: 

* We might need to have a RWmutex for the functionality struct as we are turning this on and off via GRPC. 
* A GRPC streaming endpoint to stream trade data [idea]
* A system to dynamically repair trading data holes in the event of a disconnection.

xtda commented on engine/rpcserver.go:
```diff
@@ -2238,3 +2319,294 @@ func (s *RPCServer) WebsocketSetURL(_ context.Context, r *gctrpc.WebsocketSetURL
+}
+
+// FindMissingSavedCandleIntervals is used to help determine what candle data is missing
+func (s *RPCServer) FindMissingSavedCandleIntervals(_ context.Context, r *gctrpc.FindMissingCandlePeriodsRequest) (*gctrpc.FindMissingIntervalsResponse, error) {
+	if r.End <= 0 || r.Start <= 0 || r.ExchangeName == "" || r.Pair == nil || r.AssetType == "" || r.Pair.String() == "" || r.Interval <= 0 {
+		return nil, errors.New("invalid arguments received")
+	}
+	startTime := time.Unix(r.Start, 0)
+	endTime := time.Unix(r.End, 0)
+	klineItem, err := kline.LoadFromDatabase(
```
Not sure if its worth adding an exchange enabled check on this and FindMissingSavedTradeIntervals below to only allow its usage on exchanges currently enabled by bot

xtda commented on exchanges/bitstamp/bitstamp.go:
```diff
@@ -231,21 +231,21 @@ func (b *Bitstamp) GetTradingPairs() ([]TradingPair, error) {
+		strings.ToLower(currencyPair),
 	)
+	if timePeriod != "" {
+		requestURL += "?time=" + timePeriod
```
While its unlikely anything will ever break it might be be worth wrapping timePeriod in url.QueryEscape()

xtda commented on cmd/exchange_wrapper_issues/main.go:
```diff
@@ -408,10 +409,67 @@ func testWrappers(e exchange.IBotExchange, base *exchange.Base, config *Config)
 				Response:   jsonifyInterface([]interface{}{nil}),
 			})
+
+			var getExchangeHistoryResponse []trade.Data
```
Rename to getHistoricTradesResponse to keep consistent with new wrapper name

xtda reviewed (changes requested):
The fun of having to review someone else's large PR's for a change :D 
not many changes from me just a few minor QoL ones

gloriousCode commented on cmd/gctcli/commands.go:
```diff
@@ -3775,6 +3776,10 @@ var getHistoricCandlesCommand = cli.Command{
 			Destination: &candleGranularity,
 		},
+		cli.BoolFlag{
+			Name:  "fillmissingdatawithtrades",
```
I'm not sure, "fillmissingdatawithtrades" rolls off the keyboard, much like OTPNWACFEGO 👀
... That's a good idea

gloriousCode commented on exchanges/gemini/gemini_test.go:
```diff
@@ -1106,3 +1108,37 @@ func TestResponseToOrderType(t *testing.T) {
+		t.Fatal(err)
+	}
+	var resp []trade.Data
+	tStart := time.Date(2020, 6, 6, 0, 0, 0, 0, time.UTC)
```
I'll go over these again to see whether they have rolling trade history availability or not and see if I can make a nice unified way of handling them. For this test, I'll make sure its a rolling date

xtda replied:
Gemini has mock testing enabled which wont play nice with rolling dates you might need to check for that as well and alter dates depending on that

gloriousCode replied:
So looking further into this, some exchanges don't really document their trade history limits, so its difficult to account for. Plus they can change their policy whenever. I'm going to expand testing and update processing to verify the timestamps in returned responses and exit if it doesn't meet the params. I won't have hardcoded dates in tests except to highlight that an exchange definitely can't retrieve data from that point

gloriousCode commented on database/repository/trade/trade_test.go:
```diff
@@ -0,0 +1,201 @@
+	if err != nil {
+		log.Fatal(err)
+	}
+	os.Exit(m.Run())
```
As discussed, latest golanglint-ci linters still want it there. So I've kept it

gloriousCode commented:
>Out of scope considerations:
 >  1. We might need to have a RWmutex for the functionality struct as we are turning this on and off via GRPC.
  >  2. A GRPC streaming endpoint to stream trade data [idea]
   >3. A system to dynamically repair trading data holes in the event of a disconnection.
I've done 1. I think 2 would come in its own PR.

For 3, I can think of two options. The first could be a GRPC endpoint which is like the `FindMissingSavedTradeIntervals` endpoint, but will attempt to fill the discovered missing areas. The second we could have a routine which iterates over any exchange that has `saveTradeData` enabled and call `GetRecentTrades` from the wrapper every 30 minutes. What are your thoughts?

shazbert commented:
> > Out of scope considerations:
> > 
> > 1. We might need to have a RWmutex for the functionality struct as we are turning this on and off via GRPC.
> > 2. A GRPC streaming endpoint to stream trade data [idea]
> > 3. A system to dynamically repair trading data holes in the event of a disconnection.
> 
> I've done 1. I think 2 would come in its own PR.
> 
> For 3, I can think of two options. The first could be a GRPC endpoint which is like the `FindMissingSavedTradeIntervals` endpoint, but will attempt to fill the discovered missing areas. The second we could have a routine which iterates over any exchange that has `saveTradeData` enabled and call `GetRecentTrades` from the wrapper every 30 minutes. What are your thoughts?

Those ideas were pretty much just thought bubbles. Thanks for including 1 into the PR. For number 3 I think it should be its own PR as well. Both suggestions sound good if we were to do the routine avenue, we might set up a system for checking OHLC database data as well; still thought bubble.

shazbert reviewed (changes requested):
Thanks for making changes, nothing critical, some niterinos.

gloriousCode commented on engine/rpcserver.go:
```diff
@@ -1586,7 +1586,7 @@ func (s *RPCServer) GetAuditEvent(_ context.Context, r *gctrpc.GetAuditEventRequ
 				Identifier: v[x].Identifier,
 				Message:    v[x].Message,
-				Timestamp:  v[x].CreatedAt.In(loc).Format(common.SimpleTimeFormat),
+				Timestamp:  v[x].CreatedAt.In(time.Local).Format(common.SimpleTimeFormat),
```
Interesting consideration! Would you like me to do this for all forms of timestamps for rpcserver?

shazbert replied:
Depending on the work involved you could do it as a separate PR. So force the client to handle relative conversions.

gloriousCode replied:
I've add a new time format which includes the timezone so you always know its UTC for RPC endpoint output

shazbert replied:
Epic!

shazbert commented on exchanges/coinbene/coinbene_websocket.go:
```diff
@@ -55,22 +58,36 @@ func (c *Coinbene) WsConnect() error {
 			subscriptions = append(subscriptions, stream.ChannelSubscription{
-				Channel:  fmt.Sprintf(channels[x], pairs[y]),
-				Currency: pairs[y],
+				Channel:  swapChannelPreix + fmt.Sprintf(channels[x], perpetualPairs[y]),
```
🦅

shazbert commented on cmd/gctcli/commands.go:
```diff
@@ -4019,13 +4047,238 @@ func getHistoricCandlesExtended(c *cli.Context) error {
+	Flags: []cli.Flag{
+		cli.StringFlag{
+			Name:  "exchange, e",
+			Usage: "the exchange to get the candles from",
```
candles  -> trades

shazbert commented on cmd/gctcli/commands.go:
```diff
@@ -4019,13 +4047,238 @@ func getHistoricCandlesExtended(c *cli.Context) error {
+		},
+		cli.StringFlag{
+			Name:  "pair, p",
+			Usage: "the currency pair to get the candles for",
```
candles -> trades

shazbert commented on cmd/gctcli/commands.go:
```diff
@@ -4019,13 +4047,238 @@ func getHistoricCandlesExtended(c *cli.Context) error {
+	Flags: []cli.Flag{
+		cli.StringFlag{
+			Name:  "exchange, e",
+			Usage: "the exchange to get the candles from",
```
candles -> trades

shazbert commented on cmd/gctcli/commands.go:
```diff
@@ -4019,13 +4047,238 @@ func getHistoricCandlesExtended(c *cli.Context) error {
+		},
+		cli.StringFlag{
+			Name:  "pair, p",
+			Usage: "the currency pair to get the candles for",
```
candles -> trades

shazbert commented on cmd/gctcli/commands.go:
```diff
@@ -4034,3 +4287,660 @@ func getHistoricCandlesExtended(c *cli.Context) error {
+	Flags: []cli.Flag{
+		cli.StringFlag{
+			Name:  "exchange, e",
+			Usage: "the exchange to get the candles from",
```
candles -> trades

shazbert commented on cmd/gctcli/commands.go:
```diff
@@ -4034,3 +4287,660 @@ func getHistoricCandlesExtended(c *cli.Context) error {
+		},
+		cli.StringFlag{
+			Name:  "pair, p",
+			Usage: "the currency pair to get the candles for",
```
candles -> trades

shazbert commented on cmd/gctcli/commands.go:
```diff
@@ -4034,3 +4287,660 @@ func getHistoricCandlesExtended(c *cli.Context) error {
+	Flags: []cli.Flag{
+		cli.StringFlag{
+			Name:  "exchange, e",
+			Usage: "the exchange to get the candles from",
```
candles -> trades

shazbert commented on cmd/gctcli/commands.go:
```diff
@@ -4034,3 +4287,660 @@ func getHistoricCandlesExtended(c *cli.Context) error {
+		},
+		cli.StringFlag{
+			Name:  "pair, p",
+			Usage: "the currency pair to get the candles for",
```
candles -> trades

shazbert commented on cmd/gctcli/commands.go:
```diff
@@ -4034,3 +4287,660 @@ func getHistoricCandlesExtended(c *cli.Context) error {
+	Flags: []cli.Flag{
+		cli.StringFlag{
+			Name:  "exchange, e",
+			Usage: "the exchange to get the candles from",
```
probably rethink the usage desc

shazbert reviewed (changes requested):
Testing find missing saved trade intervals:
```
go run ./... findmissingsavedtradeintervals --e="bitmex" --pair="eth-usd" --asset="perpetualcontract" --start="2020-09-17 00:00:00"
"missingPeriods": [
  "2020-09-16 13:00:00 UTC - 2020-09-16 14:00:00 UTC",
  "2020-09-16 14:00:00 UTC - 2020-09-16 15:00:00 UTC",
  "2020-09-16 15:00:00 UTC - 2020-09-16 16:00:00 UTC",
  "2020-09-16 16:00:00 UTC - 2020-09-16 17:00:00 UTC",
  "2020-09-16 17:00:00 UTC - 2020-09-16 18:00:00 UTC",
  "2020-09-16 18:00:00 UTC - 2020-09-16 19:00:00 UTC",
  "2020-09-16 19:00:00 UTC - 2020-09-16 20:00:00 UTC",
  "2020-09-16 20:00:00 UTC - 2020-09-16 21:00:00 UTC",
  "2020-09-16 21:00:00 UTC - 2020-09-16 22:00:00 UTC",
  "2020-09-16 22:00:00 UTC - 2020-09-16 23:00:00 UTC",
  "2020-09-16 23:00:00 UTC - 2020-09-17 00:00:00 UTC",
  "2020-09-17 00:00:00 UTC - 2020-09-17 01:00:00 UTC",
  "2020-09-17 01:00:00 UTC - 2020-09-17 02:00:00 UTC",
  "2020-09-17 02:00:00 UTC - 2020-09-17 03:00:00 UTC",
  "2020-09-17 03:00:00 UTC - 2020-09-17 04:00:00 UTC",
  "2020-09-17 04:00:00 UTC - 2020-09-17 05:00:00 UTC",
  "2020-09-17 07:00:00 UTC - 2020-09-17 08:00:00 UTC",
  "2020-09-17 08:00:00 UTC - 2020-09-17 09:00:00 UTC",
  "2020-09-17 09:00:00 UTC - 2020-09-17 10:00:00 UTC",
  "2020-09-17 10:00:00 UTC - 2020-09-17 11:00:00 UTC",
  "2020-09-17 11:00:00 UTC - 2020-09-17 12:00:00 UTC",
  "2020-09-17 12:00:00 UTC - 2020-09-17 13:00:00 UTC",
  "2020-09-17 13:00:00 UTC - 2020-09-17 14:00:00 UTC",
  "2020-09-17 14:00:00 UTC - 2020-09-17 15:00:00 UTC",
  "2020-09-17 15:00:00 UTC - 2020-09-17 16:00:00 UTC",
  "2020-09-17 16:00:00 UTC - 2020-09-17 17:00:00 UTC",
  "2020-09-17 17:00:00 UTC - 2020-09-17 18:00:00 UTC",
  "2020-09-17 18:00:00 UTC - 2020-09-17 19:00:00 UTC",
  "2020-09-17 19:00:00 UTC - 2020-09-17 20:00:00 UTC",
  "2020-09-17 20:00:00 UTC - 2020-09-17 21:00:00 UTC",
  "2020-09-17 21:00:00 UTC - 2020-09-17 22:00:00 UTC",
  "2020-09-17 22:00:00 UTC - 2020-09-17 23:00:00 UTC",
  "2020-09-17 23:00:00 UTC - 2020-09-18 00:00:00 UTC",
continued ...
"2020-09-28 02:00:00 UTC - 2020-09-28 03:00:00 UTC"
```

Returned:
```
"status": "Found 1960 periods. Missing 276 periods between 2020-09-28 04:00:00 UTC and 2020-09-28 03:39:16 UTC" (time here seems to be off was expecting to have "2020-09-16 13:00:00 UTC -> 2020-09-17 05:00:00 UTC" AND "2020-09-17 06:00:00 UTC - 2020-09-28 03:00:00 UTC" )
```
I currently have data from "2020-09-17 15:08:40.727+10" -> "2020-09-17 15:29:24.494+10" so this works at cutting out the "2020-09-17 05:00:00 UTC -> 2020-09-17 06:00:00 UTC" && "2020-09-17 07:00:00 UTC -> 2020-09-17 07:00:00 UTC", hourly check 🚀 

Also, two items that are not working as intended:
* gocryptotrader\cmd\gctcli> go run ./... getrecenttrades --e="bitmex" --pair="eth-usd" --asset=perpetualcontract <- This hangs and doesn't produce anything for some reason
* gocryptotrader\cmd\gctcli> go run ./... gethistorictrades --e="bitmex" --pair="eth-usd" --asset=perpetualcontract <- this as well

gloriousCode commented:
> Also, two items that are not working as intended:
>     gocryptotrader\cmd\gctcli> go run ./... getrecenttrades --e="bitmex" --pair="eth-usd" --asset=perpetualcontract <- This hangs and doesn't produce anything for some reason
>     gocryptotrader\cmd\gctcli> go run ./... gethistorictrades --e="bitmex" --pair="eth-usd" --asset=perpetualcontract <- this as well

This was actually working. There it was just taking a long time to get all the trades in that timeframe -to the point where once it did finally respond it would be larger than the allowed GRPC response size.

I've reduced the default timespan when retrieving recent trades and historic trades. I've also change the historic trade response to a stream and update the caller more frequently as a result
```
.\gctcli.exe gethistorictrades --e="bitmex" --pair="eth-usd" --asset=perpetualcontract
2:10PM  | Beginning stream retrieving trades in 1 hour batches from 2020-09-29 08:10:05 UTC to 2020-09-29 14:10:05 UTC
2:10PM  | If you have provided a large time range, please be patient
2:10PM  | Processed 2392 trades between 2020-09-28 22:10:17 UTC and 2020-09-28 23:09:33 UTC
2:12PM  | Processed 14201 trades between 2020-09-28 23:10:05 UTC and 2020-09-29 00:10:00 UTC
2:13PM  | Processed 2295 trades between 2020-09-29 00:10:06 UTC and 2020-09-29 01:09:54 UTC
2:13PM  | Processed 3039 trades between 2020-09-29 01:10:11 UTC and 2020-09-29 02:10:02 UTC
2:14PM  | Processed 1700 trades between 2020-09-29 02:10:08 UTC and 2020-09-29 03:09:52 UTC
2:14PM  | Processed 1987 trades between 2020-09-29 03:10:15 UTC and 2020-09-29 04:09:28 UTC
2:14PM  | Trade retrieval complete! Process took 4m16.2964591s
```

I haven't yet cracked missing trade period condensing though.

gloriousCode commented:
Responses are now using time ranges:
```
.\gctcli.exe trade findmissingsavedtradeintervals -e=zb --pair=btc-usdt --asset=spot -start="2020-09-10 09:47:21"
{
 "exchangeName": "zb",
 "assetType": "spot",
 "pair": {
  "delimiter": "-",
  "base": "btc",
  "quote": "usdt"
 },
 "missingPeriods": [
  "2020-09-09 23:00:00 UTC - 2020-09-15 23:00:00 UTC",
  "2020-09-16 00:00:00 UTC - 2020-10-01 02:00:00 UTC"
 ],
 "status": "Found 1 periods. Missing 2 periods between 2020-09-09 23:00:00 UTC and 2020-10-01 02:00:00 UTC"
}
```
So there's now a new timeperiods package to help with that. I also moved trade related CLI commands to `trade` subcommands

shazbert reviewed (changes requested):
This is an epic update, thanks for addressing nits! Got some basic ones that I found! 👍

xtda commented on common/timeperiods/timeperiods.go:
```diff
@@ -0,0 +1,137 @@
+)
+
+// FindTimeRangesContainingData will break the start and end into time periods using the provided period
+// it will then check whether any comparisonTimes are within those periods and concatinate them
```
concatinate -> concatenate ?

xtda commented on common/timeperiods/timeperiods_types.go:
```diff
@@ -0,0 +1,43 @@
+}
+
+// Sort will sort the time period asc or desc
+func (t *TimePeriodCalculator) Sort(desc bool) {
```
Can add test coverage to the descending condition here if you want to increase your test coverage

Also is it worth moving this out of the types file into timeperiods.go ?

xtda reviewed (changes requested):
all the other changes working great 👍 
Just two small changes and it looks good to go for me

xtda reviewed (commented):
no code issues from me 

Just one issue regarding the usage of the gctlog package in gctcli without any configuration or call to SetupGlobalLogger()

there are only two occurrences of it at the moment but without the setup mentioned above logging is disabled by default

Don't mind if you decide to use gctlog with default settings or just the default log/fmt setup (as gctlog is a little heavy for simple outputs)

xtda commented on cmd/gctcli/trades.go:
```diff
@@ -0,0 +1,756 @@
+		}
+	}()
+
+	s, err := time.Parse(common.SimpleTimeFormat, startTime)
+	if err != nil {
+		return fmt.Errorf("invalid time format for start: %v", err)
+	}
+	e, err := time.Parse(common.SimpleTimeFormat, endTime)
+	if err != nil {
+		return fmt.Errorf("invalid time format for end: %v", err)
+	}
+
```
If we move this check up before setupClient() on 255 it will save attempting to open a connection to the grpc server if timestamp parsing fails

gloriousCode replied:
:+1: Just did this for all instances I came across for all CLI commands

xtda commented on cmd/gctcli/commands.go:
```diff
@@ -4036,13 +4063,164 @@ func getHistoricCandlesExtended(c *cli.Context) error {
+
+	candleInterval := time.Duration(candleGranularity) * time.Second
+
+	s, err := time.Parse(common.SimpleTimeFormat, startTime)
+	if err != nil {
+		return fmt.Errorf("invalid time format for start: %v", err)
+	}
+	e, err := time.Parse(common.SimpleTimeFormat, endTime)
+	if err != nil {
+		return fmt.Errorf("invalid time format for end: %v", err)
+	}
+
+	if e.Before(s) {
+		return errors.New("start cannot be after end")
+	}
+
```
same as above

If we move this check up before setupClient() on 4185 it will save attempting to open a connection to the grpc server if timestamp parsing fails

xtda commented on cmd/gctcli/trades.go:
```diff
@@ -0,0 +1,756 @@
+		}
+	}()
+
+	s, err := time.Parse(common.SimpleTimeFormat, startTime)
+	if err != nil {
+		return fmt.Errorf("invalid time format for start: %v", err)
+	}
+	e, err := time.Parse(common.SimpleTimeFormat, endTime)
+	if err != nil {
+		return fmt.Errorf("invalid time format for end: %v", err)
+	}
+
+	if e.Before(s) {
+		return errors.New("start cannot be after end")
+	}
```
If we move this check up before setupClient() on 394 it will save attempting to open a connection to the grpc server if timestamp parsing fails

xtda commented on cmd/gctcli/trades.go:
```diff
@@ -0,0 +1,756 @@
+		}
+	}()
+
+	s, err := time.Parse(common.SimpleTimeFormat, startTime)
+	if err != nil {
+		return fmt.Errorf("invalid time format for start: %v", err)
+	}
+	e, err := time.Parse(common.SimpleTimeFormat, endTime)
+	if err != nil {
+		return fmt.Errorf("invalid time format for end: %v", err)
+	}
+
+	if e.Before(s) {
+		return errors.New("start cannot be after end")
+	}
```
If we move this check up before setupClient() on 563 it will save attempting to open a connection to the grpc server if timestamp parsing fails

xtda commented on cmd/gctcli/trades.go:
```diff
@@ -0,0 +1,756 @@
+
+	candleInterval := time.Duration(candleGranularity) * time.Second
+
+	s, err := time.Parse(common.SimpleTimeFormat, startTime)
+	if err != nil {
+		return fmt.Errorf("invalid time format for start: %v", err)
+	}
+	e, err := time.Parse(common.SimpleTimeFormat, endTime)
+	if err != nil {
+		return fmt.Errorf("invalid time format for end: %v", err)
+	}
+
+	if e.Before(s) {
+		return errors.New("start cannot be after end")
+	}
```
If we move this check up before setupClient() on 708 it will save attempting to open a connection to the grpc server if timestamp parsing fails

xtda reviewed (changes requested):
Last change(s?) all the same thing but just one tiny execution order change to save opening connection to server before all validation is done

xtda reviewed (approved):
new changes working good

docker still working which is always a benefit
writing to both sqlite & postgres including migrations 
retrieving data successfully now with correct time zones

Approved 🎉

shazbert reviewed (changes requested):
https://travis-ci.com/github/thrasher-corp/gocryptotrader/jobs/396661411#L551
Needs to be fixed. 
Everything else looks good!

shazbert commented on engine/rpcserver.go:
```diff
@@ -2255,3 +2339,494 @@ func (s *RPCServer) WebsocketSetURL(_ context.Context, r *gctrpc.WebsocketSetURL
+		Asset:        r.AssetType,
+		Pair:         r.Pair,
+	}
+	for i := range trades {
+		resp.Trades = append(resp.Trades, &gctrpc.SavedTrades{
+			Price:     trades[i].Price,
+			Amount:    trades[i].Amount,
+			Side:      trades[i].Side.String(),
+			Timestamp: trades[i].Timestamp.In(time.UTC).Format(common.SimpleTimeFormatWithTimezone),
+			TradeId:   trades[i].TID,
+		})
+	}
+	for i := range trades {
+		resp.Trades = append(resp.Trades, &gctrpc.SavedTrades{
+			Price:     trades[i].Price,
+			Amount:    trades[i].Amount,
+			Side:      trades[i].Side.String(),
+			Timestamp: trades[i].Timestamp.In(time.UTC).Format(common.SimpleTimeFormatWithTimezone),
+			TradeId:   trades[i].TID,
+		})
+	}
```
👀

shazbert reviewed (changes requested):
Thanks for fixing that, just going over everything once again. Cannot seem to set binance processing:
```go run ./... trade setexchangetradeprocessing binance true```
Won't save recent to database using:
```go run ./... trade getrecent --exchange=binance --pair=ltc-usdt --asset=spot```

shazbert reviewed (changes requested):
Thanks for adding an update, it was really my issue not the codes, but what you have done is awesome. Found a bug for binance exchange: 
![image](https://user-images.githubusercontent.com/9391715/95699274-c632b280-0c8f-11eb-8d68-6db1368c054c.png)
Random BUY side and timestamp inserted in line with trade ID. Could be a websocket/rest issue.

Also when returning the payload of trades via GCTCLI we are only getting side == "BUY" when it should be unknown?

Going to do another sweep across some exchanges with WebSocket enabled to see if there is any variance.

shazbert commented:
Will update this list as I find issues:
- [ ] Bitfinex v2 trades, amount +/- dictates buy or sell side in our HTTP Request (Websocket is fine)

shazbert reviewed (changes requested):
The full sweep of recent trade functionality and issues found: https://github.com/thrasher-corp/gocryptotrader/pull/558#issuecomment-706828229
Tested websocket and rest combination and deployment to database.

gloriousCode commented:
So LakeBTC's websocket trade subscription does not provide enough data to satisfy our requirements for trade processing alongside the REST endpoint. So I've dropped the websocket support of trades.

I've also updated the constraints of the database tables via a new migration. It keeps the constraint `uniquetradeid`, but this will ensure that the other unique constraint of `base,quote,asset,price,amount,side, timestamp` is only applied when the tradeid is not supplied (previously, it would conflict on match when there was a tradeid present).

shazbert reviewed (changes requested):
Changes tested, migrated up and down on sqlite3. 
- [x] Localbitcoins is good
- [x] OkExchanges good for REST deployment to DB now. 

Fine work sir just one nit for reversion, the null request on the unknown side is something I don't really mind not being changed. After this should be good to go. **Should....**

shazbert reviewed (approved):
tACK! 💯 Thanks for making those painful changes!

MadCozBadd commented on exchanges/zb/zb_websocket.go:
```diff
@@ -214,37 +215,42 @@ func (z *ZB) wsHandleData(respRaw []byte) error {
-
-			tSide, err := order.StringToOrderSide(trades.Data[i].TradeType)
+			var tSide order.Side
+			tSide, err = order.StringToOrderSide(tradeData.Data[i].TradeType)
```
They have 2 fields, tradeData.Data[i].Type and tradeData.Data[i].TradeType

Type has buy/sell (which relates to order side in GCT) however their tradetype returns bid/ask which makes no sense.
Pls double check n if everything is as intended then just ignore this comment: https://www.zb.com/api#wugiixorsxsjshm60

gloriousCode replied:
While the code does convert bid/ask to buy/sell now, I've updated this to use `Type` instead

MadCozBadd reviewed (commented):
Nothing much from me unfortunately :/

shazbert reviewed (approved):
new changes tACK

MadCozBadd reviewed (approved):
tACK!!!!!!!!!!!!!!

xtda reviewed (approved):
tested with all the new changes and fixes since the last approval

![image](https://user-images.githubusercontent.com/1223381/96190915-bfa58300-0f8e-11eb-9923-4cb802303f4a.png)

thrasher- reviewed (changes requested):
Happy little compile error after the latest master merge: https://travis-ci.com/github/thrasher-corp/gocryptotrader/jobs/400456742#L260

MadCozBadd reviewed (approved):
reTACK

thrasher- reviewed (changes requested):
Really awesome work, just some minor nits. Tested with BTC Markets and Bitfinex with PSQL:

![image](https://user-images.githubusercontent.com/4685270/96948210-ce60dc80-1530-11eb-8b0d-1b04554e5ecf.png)

Expected behavior matched with the output 👍

thrasher- commented on main.go:
```diff
@@ -85,6 +85,7 @@ func main() {
 	flag.StringVar(&settings.HTTPUserAgent, "httpuseragent", "", "sets the HTTP user agent")
 	flag.StringVar(&settings.HTTPProxy, "httpproxy", "", "sets the HTTP proxy server")
 	flag.BoolVar(&settings.EnableExchangeHTTPDebugging, "exchangehttpdebugging", false, "sets the exchanges HTTP debugging")
+	flag.DurationVar(&settings.TradeBufferProcessingInterval, "tradeprocessinginterval", time.Second*15, "sets the interval to save trade buffer data to the database")
```
Can use trade.DefaultProcessorIntervalTime, similar to how we do it with request.DefaultMaxRequestJobs

thrasher- reviewed (changes requested):
Thanks for making those changes! Only one last nit

shazbert reviewed (commented):
Looks good to me. reACK when the last nit is done.

MadCozBadd reviewed (commented):
will re tACK after the last nit changes too. tested the rest n it LGTM

thrasher- reviewed (approved):
tACK, great work!

shazbert reviewed (approved):
tACK:
![image](https://user-images.githubusercontent.com/9391715/97236217-bb078700-1838-11eb-8cd2-5b5cd71e8c06.png)

shazbert commented on exchanges/kraken/kraken.go:
```diff
@@ -350,14 +350,16 @@ func (k *Kraken) GetTrades(symbol string) ([]RecentTrades, error) {
 
-	trades, ok := tradeInfo[symbol].([]interface{})
+	var trades []interface{}
+	trades, ok = tradeInfo[symbol].([]interface{})
```
Not your issue but this can slightly obfuscate an issue where a symbol won't be matched in the lookup table and data above might panic if we get an issue. 

```
data, ok := result.(map[string]interface{})
	if !ok {
		return recentTrades, errors.New("unexpected data received")
	}

	resulty, ok := data["result"]
	if !ok {
		return recentTrades, errors.New("unexpected data received")
	}

	tradeInfo, ok := resulty.(map[string]interface{})
	if !ok {
		return recentTrades, errors.New("unexpected data received")
	}

	tradeyTrades, ok := tradeInfo[symbol]
	if err != nil {
		return recentTrades, fmt.Errorf("symbol [%s] not found in returned data", symbol)
	}

	var trades []interface{}
	trades, ok = tradeyTrades.([]interface{})
	if !ok {
		return recentTrades, errors.New("unexpected data received")
	}
```

shazbert reviewed (changes requested):
Found pedantic issue. Sorry.

shazbert reviewed (approved):
re-tACK - Tested kraken trades.

MadCozBadd reviewed (approved):
reTACK!
